content/blog/the-complete-beginners-guide-to-cli-apps-in-v.md

```diff
@@ -0,0 +1,837 @@
+		mut current_line := []rune{cap: options.size + i + 1}
+		for _ in 0 .. (options.size - i - 1) {
+			current_line << ` `
+		}
+		for _ in 0 .. (2 * i + 1) {
+			current_line << options.symbol
+		}
+		lines << current_line.string()
+	}
+
+	// bottom half
+	for i := options.size - 2; i >= 0; i-- {
+		lines << lines[i]
+	}
+	return lines
+}
+```
+
+All we are doing here is simple manipulations of arrays using nesting
+and loops with interesting indices. Sometimes, we reuse the previous
+lines and copy them over to simplify the shape generation, like we do
+in the `generate_diamond` function.
+
+In every function, we make use of `options.are_valid()` to check if
+the options are valid. If they are, we return the shape. If they are
+not, we return an empty array.
+
+Finally, we can modify the `generate_shape` function to use the appropriate
+functions and return the shapes as requested.
+
+```v
+pub fn generate_shape(options ShapeOptions) []string {
+	return match options.kind {
+		.left_triangle { generate_left_triangle(options) }
+		.right_triangle { generate_right_triangle(options) }
+		.pyramid { generate_pyramid(options) }
+		.square { generate_square(options) }
+		.diamond { generate_diamond(options) }
+	}
+}
+```
+
+Now when we run the project, we see the following outputs:
+
+```text
+$ v run .
+Enter a shape: square
+*****
+*****
+*****
+*****
+*****
+$ v run . --size 10 --shape right-triangle --symbol "/"
+         /
+        //
+       ///
+      ////
+     /////
+    //////
+   ///////
+  ////////
+ /////////
+//////////
+```
+
+## Final Touches
+
+For future maintainability, we moved the application information to the
+`geometry/metadata.v` file:
+
+```v {linenos=table}
+module geometry
+
+import v.vmod
+
+const manifest = vmod.from_file('v.mod') or { panic(err) }
+
+pub const (
+	version     = manifest.version
+	name        = manifest.name
+	description = manifest.description
+)
+```
+
+Now we are directly extracting the name, version, description from the
+`v.mod` file. So we only need to keep the information updated in one place.
+We can modify the relevant part of the `geo.v` file as well:
+
+```v
+// ...
+
+fn main() {
+	mut fp := flag.new_flag_parser(os.args)
+
+	fp.application(geometry.name)
+	fp.version(geometry.version)
+	fp.description(geometry.description)
+	fp.skip_executable()
+	// ...
+}
+```
+
+Now, when we want to bump the version number for the project, we can just do
+`v bump --patch`, `v bump --minor` or `v bump --major` at the root of the
+repository.
+
+## Final Demonstration
+
+We can now run the application with the following commands:
+
+1. `v run . --shape left-triangle --size 7 --symbol "x"`
+2. `v .` or `v -prod .` which would produce the `geo` executable.
+   Then `./geo --shape pyramid` or something similar to
+   run the final executable with the command line options.
+3. The `--help` command should also work.
+
+Some sample runs are as follows:
+
+> `v run . --shape left-triangle --size 7 --symbol "x"`
```

Comment:
This will remove the `./geo` executable after the run, and the next command will fail.
Imho it is better if you compile everything, and then only use `./geo` or, if you always use `v run .` for all sample runs.